### ISSUE TITLE
Design Specification for JUnit 5 Support

### ISSUE THREAD

**Issue by eriwen** [gradle/gradle#1037]
Produce a design spec for JUnit Platform aka JUnit 5 support in Gradle core.

Related to PR gradle/gradle#946

**Supporting information**:
[Current JUnit 5 Gradle Plugin](https://github.com/junit-team/junit5/tree/master/junit-platform-gradle-plugin/src/main/groovy/org/junit/platform/gradle/plugin)
[JUnit Javadoc](http://junit.org/junit5/docs/current/api/)
Launcher API from the [JUnit User Guide](http://junit.org/junit5/docs/current/user-guide/#launcher-api)

Expected Output: Reviewed design doc

**Comment by bmuschko** [comment]
- I merged the pull request gradle/gradle#946 though it will still need a lot of work. It was only the beginning of a spec.
- I created a little [prototype project](https://github.com/bmuschko/gradle-junit5-test) that uses the [JUnit Launcher API](http://junit.org/junit5/docs/current/user-guide/#launcher-api) from a custom task to discover and execute JUnit 5 tests to get familiar with the feature set.
- I opened a [discussion with the JUnit team](https://github.com/junit-team/junit5/issues/619) about JUnit's Java 7 compatibility and how the functionality could be leveraged in Gradle core.

**Comment by sbrannen** [comment]
@bmuschko, that's great to see that you're experimenting with a prototype!

**Comment by marcphilipp** [comment]
Any news on this issue?

**Comment by bmuschko** [comment]
@marcphilipp Thanks for checking in. Unfortunately, other priorities took over so we had to put this work on ice for a while. At the moment I see two major blockers for us:

1. [JUnit does not support Java 7](https://github.com/junit-team/junit5/issues/619). The limitation is not compatible with the Gradle core code base as is. We'd likely need to develop JUnit 5 support as a completely different module from the main build based on Java 8. I'd assume that we'll need to expose additional event listeners for resolving information we'll need for test execution and report generation.
2. We'd need to [detangle and generalize our testing support](https://github.com/gradle/gradle/pull/946#discussion_r94190127).

**Comment by marcphilipp** [comment]
Thanks for the explanation. Is there anything we can do to speed things up? Have you decided when Gradle will drop Java 7 support, yet?

**Comment by adammurdoch** [comment]
I don't think building for Java 8 is a blocker, we already do Java version specific features, and will probably continue to do so. Given this, we don't necessarily need to detangle the test support. It's a good opportunity to do so, but I don't see it as a blocker.

**Comment by bmuschko** [comment]
@adammurdoch Thanks for your input. Let's sync up on this next week to form a concrete plan forward.

**Comment by eriwen** [comment]
@marcphilipp I had a chance to catch up with Adam and @bmuschko about this. Here are some questions we need to answer so that we can move forward here:

 - [ ] We definitely want to keep parallel test execution. Does the JUnit Platform either give us control over this or have this built-in?
 - [ ] Are test execution events supported? We would at least want parity with our current events (e.g. — test started, test finished, test failed)
 - [ ] Test filtering — how are filtering by category and test name supported? 
 - [ ] Decide the appropriate level of future-proofing. We could expend a lot of effort and make this a separate project with stable public APIs, or we could use the same DSL, test filtering, event mechanisms which would not exercise the full power of the JUnit Platform but would allow us to provide support quicker.

I don't think it's on you to answer these Marc, but I'd love for us to schedule a chat so we have the information to start answering these questions. I also want to note that I expect we'll keep Java 7 support for some time, but as Adam said that isn't a blocker in our case.

Finally, I apologize for being silent here. We had to de-prioritize JUnit Platform support in the past couple of months for various reasons.

**Comment by marcphilipp** [comment]
> We definitely want to keep parallel test execution. Does the JUnit Platform either give us control over this or have this built-in?

The JUnit Platform currently does not have built-in support for parallel test execution (cf. junit-team/junit5#60). How does Gradle parallelize execution for JUnit 4? By sending separate `Requests`?

> Are test execution events supported? We would at least want parity with our current events (e.g. — test started, test finished, test failed)

Sure, there's [`TestExecutionListener`](http://junit.org/junit5/docs/current/api/org/junit/platform/launcher/TestExecutionListener.html) for that.

> Test filtering — how are filtering by category and test name supported?

Filters can be specified as part of a [LauncherDiscoveryRequest](http://junit.org/junit5/docs/current/api/org/junit/platform/launcher/LauncherDiscoveryRequest.html), e.g. there's a [TagFilter](http://junit.org/junit5/docs/current/api/org/junit/platform/launcher/TagFilter.html) (tags replace categories, you can specify the FQCN of a JUnit 4 category for backwards compatibility) and a [ClassNameFilter](http://junit.org/junit5/docs/current/api/org/junit/platform/engine/discovery/ClassNameFilter.html).

> I don't think it's on you to answer these Marc, but I'd love for us to schedule a chat so we have the information to start answering these questions.

Sure, when would be a good time for you? I have some spare time tomorrow afternoon/night (CET).

**Comment by marcphilipp** [comment]
Any news on this issue?

**Comment by eriwen** [comment]
@marcphilipp Thanks again for the ping, and sorry for the delayed response. I am asking the team to consider this as a next priority after we get Java 9 runtime support. We'll be deciding soon and if approved work can begin after Gradle 4.0 GA. Will let you know as soon as that happens.

**Comment by sormuras** [comment]
We're about to release JUnit 5 soon. Any news on this issue?

**Comment by eriwen** [comment]
We were not able to prioritize JUnit 5 support in Gradle 4.x thus far. 

@sormuras I'm sorry for not responding to this issue sooner. 

Here is the current prioritization of Epics according to our [ZenHub](https://www.zenhub.com/) board for our Core team (~2 people).

![screen shot 2017-09-05 at 3 13 00 pm](https://user-images.githubusercontent.com/51534/30085747-d7b7143a-924c-11e7-9a58-dcc9fed2bc6a.png)

This can change, but I'd say that we currently won't be able to implement JUnit Platform support earlier than Gradle 4.4 (~December 2017). This is due to other things coming up and a rough estimate of work (it will require significant rework of our testing bits).

Thank you for the follow-up. I'm sorry this may not have been the answer you're hoping for. Please reach out with further questions.

**Comment by mkobit** [comment]
Now that JUnit 5 has GA , are there any next steps for the community, or is it waiting until the Gradle team can prioritize support?

Sidenote, there was already some work on a design spec merged in https://github.com/gradle/gradle/pull/946 but it looks like the whole folder was removed from `master`.

**Comment by Lizard** [comment]
I'd like to second mkobit's question and want to add that from my company's end-user perspective, not having core Gradle support for JUnit 5 at the time of its release really is a bit of a nuisance, given the importance of the JUnit framework for Java development - I certainly hope that no further re-priorization of this feature will be necessary.

**Comment by JLLeitschuh** [comment]
@Lizard The Junit 5 plugin that the Junit 5 tesam offers is really solid. The only integrations that it's missing right now is the buildscan integrations.

**Comment by jorn86** [comment]
@JLLeitschuh Integration with source sets other than `main` is also not in that plugin. And the buildscan integrations (if I understand correctly) are important for IDE integration. So while I appreciate the JUnit team providing a basic integration plugin, this really needs to be integrated into Gradle.

**Comment by Lizard** [comment]
@JLLeitschuh Thanks for your feedback. I'm afraid that according to https://github.com/junit-team/junit5-samples/issues/35, exactly the integrations that I am looking for are currently missing and seem to depend on the Gradle core support. To quote the JUnit 5 developers: " [The Gradle] plugin is very simple and has many limitations, e.g. it does not use the test task. The plugin was meant as a proof-of-concept, not a long-term solution. Instead, we would like to start the discussion about how to get JUnit 5 support into core Gradle." (source: https://discuss.gradle.org/t/core-support-for-junit-platform-a-k-a-junit-5/19487). I'm afraid that at the moment, their plugin does not solve our problems.

**Comment by nskvortsov** [comment]
@JLLeitschuh the JUnit 5 plugin offered by JUnit team totally ignores existing test events model in Gradle. That renders IDE and CI servers integration useless.

**Comment by JLLeitschuh** [comment]
> And the buildscan integrations (if I understand correctly) are important for IDE integration.

I don't think this has anything to do with IDE integration. The buildscan is the thing published to the gradle server with reporting build performance/ect..

> the JUnit 5 plugin offered by JUnit team totally ignores existing test events model in Gradle. That renders IDE and CI servers integration useless.

I don't understand, maybe I'm not familiar with the event model.
For IDE support are you talking about executing Junit 5 tests with gradle and seeing those results show up in your IDE? Is that a thing you can do? I've always run my Junit tests without going through gradle.


I'm not advocating that Gradle not have integrations for Junit 5. I'm just saying that as a stopgap measure, the current features of the Junit 5 plugin have been adequate for myself.

I didn't know that the Gradle test task did so much. Is there a listing of the features that the Junit 5 integration needs to make this work for developers?

**Comment by nskvortsov** [comment]
>I don't understand, maybe I'm not familiar with the event model.

One can attach a [TestListener](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/testing/TestListener.html) or [TestOutputListener](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/testing/TestOutputListener.html) for a [single test task](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.testing.Test.html#org.gradle.api.tasks.testing.Test:addTestListener(org.gradle.api.tasks.testing.TestListener)) or [globally](https://docs.gradle.org/current/dsl/org.gradle.api.invocation.Gradle.html#org.gradle.api.invocation.Gradle:addListener(java.lang.Object)). This allows ide/ci to receive and handle test events and data on-the-fly.

> executing Junit 5 tests with gradle and seeing those results show up in your IDE? Is that a thing you can do?

Yes, you can (at least in IntelliJ). It is very useful when running a test also requires running other tasks (think of some kind of setup or code generation) defined in Gradle.

Overall, it feels like `junit-platform-gradle-plugin` ignores existing Gradle practices. It has separate configuration DSL (instead of extending the `test` DSL), uses `JavaExec` task type (instead of `Test` task type) and misses Gradle's core events described above. As if neither Gradle nor JUnit team is willing to take the burden of providing proper integration.

**Comment by Lizard** [comment]
@nskvortsov 
>  As if neither Gradle nor JUnit team is willing to take the burden of providing proper integration.

From reading this discussion, I actually get the feeling that the JUnit guys have been quite open and willing to do their part, but simply were never requested: https://discuss.gradle.org/t/core-support-for-junit-platform-a-k-a-junit-5/19487 

However, I think pointing fingers is a bit beside the point. For my part, I am just looking forward to getting proper JUnit 5 support into Gradle and using both how they are meant to be used :)

**Comment by nskvortsov** [comment]
@Lizard 
>using both how they are meant to be used
That would be awesome!

**Comment by jorn86** [comment]
>I don't think this has anything to do with IDE integration. The buildscan is the thing published to the gradle server with reporting build performance/ect..

I'm sorry. I was referring to the system which allows IDEs to pick up events sent from JUnit in order to display a nice overview. Apparently that's a different system then. Anyway, still important :)

**Comment by semoro** [comment]
Please, implement the full set of selectors instead of limited to single pattern `--tests` way.

Like it can be passed in [CLI Runner of JUnit](http://junit.org/junit5/docs/current/user-guide/#running-tests-console-launcher-options)

This required for launching tests fast and predictable form CLI. 

And from Gradle Tooling API 
Like in [JUnit Gradle plugin](http://junit.org/junit5/docs/current/user-guide/#running-tests-build-gradle-selectors)

This required to make possible IDE(Such as IntelliJ) select tests properly.

**Comment by serandel** [comment]
Do we have an (at least approximate) release date on this? I'm baffled that JUnit 5 and Gradle don't play well together, being both so widely used.

I would gladly donate some cash to a crowdfunding if it helped to prioritize this.

**Comment by edeandrea** [comment]
I too am voting for native support, especially being able to separate tests between source sets (i.e. test/integTest/funcTest). Having a single execution point for this as provided by the JUnit plugin doesn't solve our company's needs.

**Comment by jorn86** [comment]
Can we get an update on this? Last info was "not earlier than Gradle 4.4 (~December 2017)", and this doesn't seem to have made it into 4.4.

**Comment by eriwen** [comment]
Yes, I shared an [update on the JUnit 5 Epic here](https://github.com/gradle/gradle/issues/828#issuecomment-351431932).

**Comment by oehme** [comment]
Closing as we now have a good idea of how we want to go forward: Built-in support for JUnit 5, scheduled for Gradle 4.6. See the update on the [epic](https://github.com/gradle/gradle/issues/828#issuecomment-351431932).

**Comment by mkobit** [comment]
I know Gradle 4.6 is coming out with support for JUnit 5, but I wanted to ask a question on this issue since it seems related to one of @ajoberstar goals from the design document and the comment (https://github.com/gradle/gradle/pull/946#discussion_r94190127) - how close is the test platform/framework part of Gradle to being pluggable with other test frameworks? For example, if I wanted to plug in `pytest` or `nose` for Python testing?

**Comment by oehme** [comment]
We've learned a few lessons from JUnit 5 as well as some native test frameworks, so it's getting more generic. But making it public is not a priority yet. One thing that would help is if you tried out integrating with the internal API to find limitations for the framework you had in mind. We could use that information to drive public API decisions.

**Comment by marcphilipp** [comment]
Another possible integration path would be to write a JUnit `TestEngine` for `pytest` or `nose` and use Gradle's `useJUnitPlatform()` support. I assume for Python based testing frameworks, you'd like to specify a set of `*.py` files or directories containing `*.py` files. The JUnit Platform API supports files and directories via `FileSelector` and `DirectorySelector`, respectively. However, AFAIK all current Gradle APIs rely on Java classes.

**Comment by JLLeitschuh** [comment]
This is the only plugin I've been able to successfully use with python code/testing.
https://github.com/linkedin/pygradle

**Comment by danwallach** [comment]
Kudos where kudos are due: I adopted JUnit5 because I liked its new features, but their Gradle plugin had ... issues. Today, I upgraded from Gradle 4.5 to 4.6rc2 and made the appropriate edits in my `build.gradle` file (deleting the `org.junit.platform.gradle.plugin`, adding the call to `useJUnitPlatform()`) and... everything worked properly! The tests ran. JaCoCo is working again. Life is good. Thanks!
